### PR TITLE
runtime: remove yaml from the critical path

### PR DESCRIPTION
**Pull request by alyssawilk**
Avoiding a yaml translation step when loading runtime .
This also allows turning up e2e !yaml testing for Envoy Mobile (with limited API calls) so adding in regression testing via CI as well.

Risk Level: high
Testing: existing tests, new unit tests
Docs Changes: n/a
Release Notes: n/a
part of https://github.com/envoyproxy/envoy/issues/26450

### CODE REVIEW

File: source/common/runtime/runtime_impl.cc [hunk 1]
```diff
@@ -233,10 +233,53 @@ SnapshotImpl::Entry SnapshotImpl::createEntry(const std::string& value) {
...
+  }
+
+  entry.fractional_percent_value_ = percent;
+}
+
 SnapshotImpl::Entry SnapshotImpl::createEntry(const ProtobufWkt::Value& value) {
-  // This isn't the smartest way to do it; we're round-tripping via YAML, this should be optimized
-  // if runtime parsing becomes performance sensitive.
-  return createEntry(MessageUtil::getYamlStringFromMessage(value, false, false));
+  Entry entry;
```
Comment (htuch) [inline source/common/runtime/runtime_impl.cc:263]: Can we just make this `getJsonStringFromMessage` instead? Rationale is:
* This parsing and conversion is all fiddly and I'd probably want to see a bunch of unit tests, I don't think it's worth writing them and maintaining it unless we suspect performance really matters.
* In terms of getting rid of YAML dependency, we can use JSON here instead, since JSON is always a valid YAML subset, so the string parsing should still work.

Comment (htuch) [reply source/common/runtime/runtime_impl.cc:263]: Sorry, I missed the bit in the linked issue referencing JSON. Can we at least factor these utlls into a library in common/common and add tests then?

Comment (alyssawilk) [reply source/common/runtime/runtime_impl.cc:263]: we can, but they're only used by runtime, and covered by the existing runtime unit tests, in particular https://github.com/envoyproxy/envoy/blob/main/test/common/runtime/runtime_impl_test.cc#L582 which I used to make sure invalid input was still handled the same way it was previously.

Comment (alyssawilk) [reply source/common/runtime/runtime_impl.cc:263]: actually swapping reviews as htuch mentioned he was ooo

File: source/common/runtime/runtime_impl.cc [hunk 2]
```diff
@@ -233,10 +233,53 @@ SnapshotImpl::Entry SnapshotImpl::createEntry(const std::string& value) {
...
+    if (f.first == "numerator") {
+      if (f.second.has_number_value()) {
+        percent.set_numerator(f.second.number_value());
+      }
+    } else if (f.first == "denominator" && f.second.has_string_value()) {
+      if (f.second.string_value() == "HUNDRED") {
+        percent.set_denominator(envoy::type::v3::FractionalPercent::HUNDRED);
+      } else if (f.second.string_value() == "TEN_THOUSAND") {
+        percent.set_denominator(envoy::type::v3::FractionalPercent::TEN_THOUSAND);
+      } else if (f.second.string_value() == "MILLION") {
```
Comment (ggreenway) [inline source/common/runtime/runtime_impl.cc:249]: Add a `static_assert(envoy::type::v3::FractionalPercent::MILLION == envoy::type::v3::FractionalPercent::DenominatorType_MAX)` (or ENVOY_BUG if the compiler won't accept it)  to catch a mismatch to the proto if another value is added.

File: source/common/runtime/runtime_impl.cc [hunk 3]
```diff
@@ -233,10 +233,53 @@ SnapshotImpl::Entry SnapshotImpl::createEntry(const std::string& value) {
...
 SnapshotImpl::Entry SnapshotImpl::createEntry(const ProtobufWkt::Value& value) {
-  // This isn't the smartest way to do it; we're round-tripping via YAML, this should be optimized
-  // if runtime parsing becomes performance sensitive.
-  return createEntry(MessageUtil::getYamlStringFromMessage(value, false, false));
+  Entry entry;
+  switch (value.kind_case()) {
+  case ProtobufWkt::Value::kNumberValue:
+    entry.double_value_ = value.number_value();
+    entry.bool_value_ = value.number_value() != 0;
+    if (entry.double_value_ >= 0 && entry.double_value_ <= std::numeric_limits<uint64_t>::max()) {
```
Comment (ggreenway) [inline source/common/runtime/runtime_impl.cc:268]: This is a duplicate of code in `resolveEntryType()`. Is it worth making a helper for it? If not, copy the comment from the other one?

File: source/common/runtime/runtime_impl.cc [hunk 4]
```diff
@@ -233,10 +233,53 @@ SnapshotImpl::Entry SnapshotImpl::createEntry(const std::string& value) {
...
+      entry.uint_value_ = entry.double_value_;
+    }
+    break;
+  case ProtobufWkt::Value::kBoolValue:
+    entry.bool_value_ = value.bool_value();
+    return entry;
```
Comment (ggreenway) [inline source/common/runtime/runtime_impl.cc:274]: This case returns, all the others break. I think it's functionally the same, but inconsistent.

File: source/common/runtime/runtime_impl.cc [hunk 5]
```diff
@@ -233,10 +233,53 @@ SnapshotImpl::Entry SnapshotImpl::createEntry(const std::string& value) {
...
+  for (const auto& f : value.fields()) {
+    if (f.first == "numerator") {
+      if (f.second.has_number_value()) {
+        percent.set_numerator(f.second.number_value());
+      }
+    } else if (f.first == "denominator" && f.second.has_string_value()) {
+      if (f.second.string_value() == "HUNDRED") {
+        percent.set_denominator(envoy::type::v3::FractionalPercent::HUNDRED);
+      } else if (f.second.string_value() == "TEN_THOUSAND") {
+        percent.set_denominator(envoy::type::v3::FractionalPercent::TEN_THOUSAND);
```
Comment (ggreenway) [inline source/common/runtime/runtime_impl.cc:248]: add test coverage for this case

File: source/common/runtime/runtime_impl.cc [hunk 6]
```diff
@@ -233,10 +233,53 @@ SnapshotImpl::Entry SnapshotImpl::createEntry(const std::string& value) {
...
+      } else if (f.second.string_value() == "MILLION") {
+        percent.set_denominator(envoy::type::v3::FractionalPercent::MILLION);
```
Comment (ggreenway) [inline source/common/runtime/runtime_impl.cc:250]: add test coverage for this case

File: source/common/runtime/runtime_impl.cc [hunk 7]
```diff
@@ -233,10 +233,53 @@ SnapshotImpl::Entry SnapshotImpl::createEntry(const std::string& value) {
...
+    }
+    break;
+  case ProtobufWkt::Value::kBoolValue:
+    entry.bool_value_ = value.bool_value();
+    return entry;
+  case ProtobufWkt::Value::kStructValue:
+    parseFractionValue(entry, value.struct_value());
+    break;
+  default:
+    break;
```
Comment (ggreenway) [inline source/common/runtime/runtime_impl.cc:279]: add test coverage for this case

**Review by ggreenway — CHANGES_REQUESTED** [review]
Parsing protos manually isn't fun, but removing yaml/json from mobile is a reasonable goal, so I think this has to be done.

/wait

**Review by alyssawilk — COMMENTED** [review]
changes done.  I'm fine merging this one after the release FWIW

File: source/common/runtime/runtime_impl.cc (hunk 3 above)
Comment (alyssawilk) [reply source/common/runtime/runtime_impl.cc:270]: reworked so I can remove the yaml conversion entirely (except from file and admin ingress, where it's needed, and I'll handle separately).  This gets us significantly more test coverage including some wonky corner cases (like shoving a yaml struct into a proto string) which I added a TODO to deprecate if we can.

**Comment by alyssawilk** [comment]
(still don't want to merge until release but I'd be happy to progress on review today/tomorrow)

**Comment by RyanTheOptimist** [comment]
Looks like a real failure: https://dev.azure.com/cncf/envoy/_build/results?buildId=134041&view=logs&jobId=71bbc8ed-7425-56af-5e60-50f314c92f4d&j=71bbc8ed-7425-56af-5e60-50f314c92f4d&t=d13ab329-77cb-5916-e304-571c2e9a8427

```
source/common/runtime/runtime_impl.cc:255:33: runtime error: 1.84467e+20 is outside the range of representable values of type 'int'
    #0 0x79a585d in Envoy::Runtime::setNumberValue(Envoy::Runtime::Snapshot::Entry&, double) /proc/self/cwd/source/common/runtime/runtime_impl.cc:255:33
    #1 0x79a6aa2 in Envoy::Runtime::SnapshotImpl::createEntry(google::protobuf::Value const&, std::__1::basic_string_view<char, std::__1::char_traits<char> >) /proc/self/cwd/source/common/runtime/runtime_impl.cc:270:5
    #2 0x79a8d41 in Envoy::Runtime::DiskLayer::walkDirectory(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned int, Envoy::Api::Api&) /proc/self/cwd/source/common/runtime/runtime_impl.cc:403:25
    #3 0x79a777e in Envoy::Runtime::DiskLayer::DiskLayer(std::__1::basic_string_view<char, std::__1::char_traits<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, Envoy::Api::Api&) /proc/self/cwd/source/common/runtime/runtime_impl.cc:347:3
    #4 0x79c0855 in std::__1::__unique_if<Envoy::Runtime::DiskLayer>::__unique_single std::__1::make_unique<Envoy::Runtime::DiskLayer, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&, Envoy::Api::Api&>(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&, Envoy::Api::Api&) /opt/llvm/bin/../include/c++/v1/__memory/unique_ptr.h:725:32
    #5 0x79b597b in Envoy::Runtime::LoaderImpl::createNewSnapshot() /proc/self/cwd/source/common/runtime/runtime_impl.cc:658:31
    #6 0x79aebbb in Envoy::Runtime::LoaderImpl::loadNewSnapshot() /proc/self/cwd/source/common/runtime/runtime_impl.cc:591:39
    #7 0x79ad779 in Envoy::Runtime::LoaderImpl::LoaderImpl(Envoy::Event::Dispatcher&, Envoy::ThreadLocal::SlotAllocator&, envoy::config::bootstrap::v3::LayeredRuntime const&, Envoy::LocalInfo::LocalInfo const&, Envoy::Stats::Store&, Envoy::Random::RandomGenerator&, Envoy::ProtobufMessage::ValidationVisitor&, Envoy::Api::Api&) /proc/self/cwd/source/common/runtime/runtime_impl.cc:491:3
    #8 0x3588eea in std::__1::__unique_if<Envoy::Runtime::LoaderImpl>::__unique_single std::__1::make_unique<Envoy::Runtime::LoaderImpl, Envoy::Event::MockDispatcher&, testing::NiceMock<Envoy::ThreadLocal::MockInstance>&, envoy::config::bootstrap::v3::LayeredRuntime&, testing::NiceMock<Envoy::LocalInfo::MockLocalInfo>&, Envoy::Stats::TestUtil::TestStore&, Envoy::Random::MockRandomGenerator&, testing::NiceMock<Envoy::ProtobufMessage::MockValidationVisitor>&, Envoy::Api::Api&>(Envoy::Event::MockDispatcher&, testing::NiceMock<Envoy::ThreadLocal::MockInstance>&, envoy::config::bootstrap::v3::LayeredRuntime&, testing::NiceMock<Envoy::LocalInfo::MockLocalInfo>&, Envoy::Stats::TestUtil::TestStore&, Envoy::Random::MockRandomGenerator&, testing::NiceMock<Envoy::ProtobufMessage::MockValidationVisitor>&, Envoy::Api::Api&) /opt/llvm/bin/../include/c++/v1/__memory/unique_ptr.h:725:32
    #9 0x34f2363 in Envoy::Runtime::(anonymous namespace)::DiskLoaderImplTest::run(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) /proc/self/cwd/test/common/runtime/runtime_impl_test.cc:97:15
    #10 0x34eeb07 in Envoy::Runtime::(anonymous namespace)::DiskLoaderImplTest_EmptyKeyTest_Test::TestBody() /proc/self/cwd/test/common/runtime/runtime_impl_test.cc:115:3
    #11 0xa8b5d44 in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2580:10
    #12 0xa87d5d2 in testing::Test::Run() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2655:5
    #13 0xa87f137 in testing::TestInfo::Run() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2832:11
    #14 0xa8814d4 in testing::TestSuite::Run() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2986:28
    #15 0xa8a388a in testing::internal::UnitTestImpl::RunAllTests() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:5697:44
    #16 0xa8b9fd3 in bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:2580:10
    #17 0xa8a2596 in testing::UnitTest::Run() /proc/self/cwd/external/com_google_googletest/googletest/src/gtest.cc:5280:10
    #18 0x73a59c6 in Envoy::TestRunner::RunTests(int, char**) /proc/self/cwd/external/com_google_googletest/googletest/include/gtest/gtest.h:2485:46
    #19 0x73a1116 in main /proc/self/cwd/test/main.cc:34:10
    #20 0x7f1dea7a6082 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24082) (BuildId: 1878e6b475720c7c51969e69ab2d276fae6d1dee)
    #21 0x342f32d in _start (/b/f/w/bazel-out/k8-dbg/bin/test/common/runtime/runtime_impl_test.runfiles/envoy/test/common/runtime/runtime_impl_test+0x342f32d)


```

**Review by RyanTheOptimist — DISMISSED** [review]

**Comment by phlax** [comment]
i kicked the format ci as i was watching to see impact of new caching - the error log is awful, but i think unrelated to those changes - im guessing a yaml fail of some sort

**Comment by alyssawilk** [comment]
/wait
 apps seem broken on main, will debug

File: .github/workflows/mobile-compile_time_options.yml [hunk 8]
```diff
@@ -112,4 +112,14 @@ jobs:
           --define=google_grpc=disabled \
           --define=envoy_yaml=disabled \
           --@com_envoyproxy_protoc_gen_validate//bazel:template-flavor= \
-          //:android_dist
+      //:android_dist && ./bazelw test \
```
Comment (ggreenway) [inline .github/workflows/mobile-compile_time_options.yml:115]: This seems unrelated. Was it added on accident?

**Review by ggreenway — CHANGES_REQUESTED** [review]
Aside from unrelated changes, lgtm

/wait

File: .github/workflows/mobile-compile_time_options.yml (hunk 8 above)
Comment (alyssawilk) [reply .github/workflows/mobile-compile_time_options.yml:115]: sorry related, the runtime fixes were the only thing stopping us from having e2e !yaml tests for E-M. I can do that in a follow-up PR but if there were problems I'd have to roll back together so wanted to land it together.  I'll update PR description for clarity

Comment (ggreenway) [reply .github/workflows/mobile-compile_time_options.yml:115]: Makes sense, sgtm, leave it in this PR

**Review by ggreenway — APPROVED** [review]
